AStarAlgorithmsProject/Alg Scripts/AStarSolver.cs

```diff
@@ -0,0 +1,231 @@
+            private bool passable; // if this tile on the map cannot be move to or through, this is false
+            private Point location; // the tiles location in the 2d map array
+            private TileStates state; // instance of the enum TileStates to indicate if it's been closed, open, or unchecked by the algorithim
+
+            //Default constructor, most of the time, only the minimum point should be provided, and all other parameters left to their default value
+            public Tile(Point l, bool p = true, TileStates t = TileStates.Unchecked, int cS = 1)
+            {
+                location = l;
+                passable = p;
+                state = t;
+                costScalar = cS;
+            }
+
+            //Accessor Methods
+            public bool Passable
+            { get { return passable; } set { passable = value; } }
+            public Point Location
+            { get { return location; } }
+            public TileStates State
+            { get { return state; } set{ state = value; } }
+            public int CostScalar
+            { get { return costScalar; } set { costScalar = value; } }
+
+            //Tile Heuristics 
+            public double Distance2Goal // the "h(x)" value of the heuristics. Or "cost to get there"
+            { get { return Point.Distance(location, Tile.Goal.location); } }
+            public double Distance2Start // the "g(x)" value of the huristics. Or "cost to get here"
+            { get { return Point.Distance(location, Tile.Start.location); } }
+            public double Cost // the "f(x)" value of the heuristic. Or "cost of path"
+            { get { return Distance2Start*costScalar + Distance2Goal; } }
+
+        }
+
+        private class SortTileByF : IComparer<Tile>
+        {
+            public int Compare(Tile t1, Tile t2) // if you want to have some fun, swap these values so that A* will find the most direct but least optimal solution.
+            {
+                if (t1.Cost > t2.Cost)
+                {
+                    return 1;
+                }
+                else if (t1.Cost < t2.Cost)
+                {
+                    return -1;
+                }
+                else
+                {
+                    return 0;
+                }
+            }
+        }  // Simple class to be used by the Sort method of the List class. 
+
+        public bool ValidTile(int x, int y)
+        {
+            return (x >= 0 && x < AStarSolver.MapSize) && (y >= 0 && y < AStarSolver.MapSize);
+        } // returns true if the tile at the given location is within the bounds of the map
+
+        public bool ValidTile(Point p)
+        {
+            return (p.X >= 0 && p.X < AStarSolver.MapSize) && (p.Y >= 0 && p.Y < AStarSolver.MapSize);
+        } // overload of ValidTile to use a passed point parameter
+
+        //Builds the traversal path bassed on the most ideal canidate of the current tile
+        private void TracePath() 
```

Comment:
It may improve readability if this method only returns valid neighbors and the body of the algorithm Solve method has the advanced logic.

Reply:
My take on it was we can see the Heuristic logic and the framework of what we are checking encapsulated in a single method. Regardless, I have done what you asked.